src/api/routes/rosetta/construction.ts

```diff
@@ -120,5 +131,100 @@ export function createRosettaConstructionRouter(db: DataStore): RouterWithAsync
     res.json(rosettaPreprocessResponse);
   });
 
+  //construction/metadata endpoint
+  router.postAsync('/metadata', async (req, res) => {
+    const valid: ValidSchema = await rosettaValidateRequest(req.originalUrl, req.body);
+    if (!valid.valid) {
+      res.status(400).json(makeRosettaError(valid));
+      return;
+    }
+
+    const options: RosettaOptions = req.body.options;
+    if (options.type != 'token_transfer') {
+      res.status(400).json(RosettaErrors.invalidTransactionType);
+      return;
+    }
+
+    if (options?.sender_address && !isValidC32Address(options.sender_address)) {
+      res.status(400).json(RosettaErrors.invalidSender);
+      return;
+    }
+    if (options?.symbol !== RosettaConstants.symbol) {
+      res.status(400).json(RosettaErrors.invalidCurrencySymbol);
+      return;
+    }
+
+    const recipientAddress = options.token_transfer_recipient_address;
+    if (options?.decimals !== RosettaConstants.decimals) {
+      res.status(400).json(RosettaErrors.invalidCurrencyDecimals);
+      return;
+    }
+
+    if (recipientAddress == null || !isValidC32Address(recipientAddress)) {
+      res.status(400).json(RosettaErrors.invalidRecipient);
+      return;
+    }
+
+    const accountInfo = await new StacksCoreRpcClient().getAccount(recipientAddress);
+    const nonce = accountInfo.nonce;
+
+    let recentBlockHash = undefined;
+    const blockQuery: FoundOrNot<DbBlock> = await db.getCurrentBlock();
+    if (blockQuery.found) {
+      recentBlockHash = blockQuery.result.block_hash;
+    }
+
+    const response: RosettaConstructionMetadataResponse = {
+      metadata: {
+        ...req.body.options,
+        account_sequence: nonce,
+        recent_block_hash: recentBlockHash,
+      },
+    };
+
+    res.json(response);
+  });
+
+  //construction/hash endpoint
+  router.postAsync('/hash', async (req, res) => {
+    const valid: ValidSchema = await rosettaValidateRequest(req.originalUrl, req.body);
+    if (!valid.valid) {
+      res.status(400).json(makeRosettaError(valid));
+      return;
+    }
+
+    const request: RosettaConstructionHashRequest = req.body;
+
+    let buffer: Buffer;
+    try {
+      buffer = hexToBuffer(request.signed_transaction);
+    } catch (error) {
+      res.status(400).json(RosettaErrors.invalidTransactionString);
+      return;
+    }
+
+    const transaction = deserializeTransaction(BufferReader.fromBuffer(buffer));
+    const hash = transaction.txid();
+    if (
+      transaction.auth.spendingCondition?.hashMode == AddressHashMode.SerializeP2PKH ||
+      transaction.auth.spendingCondition?.hashMode == AddressHashMode.SerializeP2WPKH
+    ) {
+      const signature = transaction.auth.spendingCondition.signature;
+      if (parseInt(signature.data, 16) == 0) {
```

Comment:
This is testing for loose equality, and it's not very clear what this trying to check for

Reply:
This is trying to check whether the transaction is signed or not. If the signature data is empty i.e `0x00` or  `0`, the transaction is not signed if it is non zero it is signed.

Reply:
But yeah I agree it is loose equality but I couldn't found any other way to differentiate between signed and unsigned transactions.

Reply:
Can it be made into an explicit check without loose equality? If the stacks-transactions-js lib changes this field, I'd prefer something to break in an explicit, clear way.